splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+                        args.append(untrusted_args)
+                elif self.cache_nonce_regex.fullmatch(untrusted_arg) \
+                        and not cache_nonce_added:
+                    args.append(untrusted_arg)
+                    cache_nonce_added = True
+                else:
+                    raise Filtered
+
+        args = b' '.join(args)
+        await self.agent_protocol.send_command(b'GENKEY', args)
+
+    async def command_SIGKEY(self, untrusted_args: Optional[bytes]):
+        args = self.verify_keygrip_arguments(1, 1, untrusted_args)
+        await self.agent_protocol.send_command(b'SIGKEY', args)
+
+    async def command_SETKEY(self, untrusted_args: Optional[bytes]):
+        args = self.verify_keygrip_arguments(1, 1, untrusted_args)
+        await self.agent_protocol.send_command(b'SETKEY', args)
+
+    async def command_SETKEYDESC(self, untrusted_args: Optional[bytes]):
+        # XXX: is there a better way than showing the message
+        #      from the untrusted domain
+        args = self.sanitize_key_desc(untrusted_args)
+        await self.agent_protocol.send_command(b'SETKEYDESC', args)
+
+    async def command_PKDECRYPT(self, untrusted_args: Optional[bytes]):
+        if untrusted_args:
+            raise Filtered
+        self.request_timer('PKDECRYPT')
+        await self.agent_protocol.send_command(b'PKDECRYPT', None)
+
+    async def command_SETHASH(self, untrusted_args: Optional[bytes]):
+        untrusted_alg, untrusted_hash = untrusted_args.split(b' ', 1)
+        try:
+            alg = int(untrusted_alg)
+            alg_param = self.hash_algos[alg]
+        except (KeyError, ValueError):
+            raise Filtered
+
+        if not untrusted_hash:
+            raise Filtered
+
+        hash_regex = re.compile(rb'[0-9A-F]{%d}' % alg_param.len)
+        if hash_regex.fullmatch(untrusted_hash) is None:
+            raise Filtered
+        hash_value = untrusted_hash
+
+        await self.agent_protocol.send_command(
+            b'SETHASH', b'%d %s' % (alg, hash_value))
+
+    async def command_PKSIGN(self, untrusted_args: Optional[bytes]):
+        if untrusted_args:
+            if untrusted_args.startswith(b'-- '):
+                untrusted_args = untrusted_args[3:]
+            if self.cache_nonce_regex.fullmatch(untrusted_args) is None:
+                raise Filtered
+            args = b'-- ' + untrusted_args
+        else:
+            args = None
+
+        self.request_timer('PKSIGN')
+
+        await self.agent_protocol.send_command(b'PKSIGN', args)
+
+    async def command_GETINFO(self, untrusted_args: Optional[bytes]):
+        if untrusted_args != b'version':
+            raise Filtered
+        args = untrusted_args
+
+        await self.agent_protocol.send_command(b'GETINFO', args)
+
+    async def command_BYE(self, untrusted_args: Optional[bytes]):
+        if untrusted_args:
+            raise Filtered
+        await self.agent_protocol.send_command(b'BYE', None)
```

Comment:
We should enforce end of communication here. Close the IO like in Ruby or something similar.

Reply:
The connection will be terminated when local agent will end it. Which it should do after receiving `BYE` command. But I can add explicit close too.

Reply:
That's true. Still fells safer. And it's easy.